Update CMSIS-Core and RTX to version 5

bulislaw opened the pull request:
## Description
Main changes made by this PR:
- Update CMSIS to v5
- Update CMSIS-RTOS to v2
- Update RTX to v5
- Remove `NVIC_SetVector` and `NVIC_GetVector` from target code as it's provided by CMSIS
- Copy interrupt vector table ROM->RAM
- Rework boot sequence
- Add RTOS1 compatibility including CMSIS adapter and RTOS1 types for both C and C++ APIs
- Optimize memory usage in tests
- Redefine thread/main/isr stack sizes
- Remove Cortex A and ARM7 support

## Status
**IN DEVELOPMENT**

## Related PRs
List related PRs against other branches:
* PAL: https://github.com/ARMmbed/mbed-client-pal/pull/75
* MCR20 driver (mesh): https://github.com/ARMmbed/mcr20a-rf-driver/pull/9
* Docs: https://github.com/ARMmbed/Handbook/pull/160

## Todos
- [ ] Fix Oulu CI (PAL: https://github.com/ARMmbed/mbed-client-pal/pull/75)
- [x] Merge work in progress work from Martin 
- [x] Move docs to handbook repo (docs/cmsis_rtx.md)
- [x] Fix failing uVisor CI
- [x] Revert changes that make Nanostack examples pass (PI name) and fix it properly
- [x] Fix mesh minimal tests failing by merging https://github.com/ARMmbed/mcr20a-rf-driver/pull/9

bulislaw committed:
CMSIS5: Replace target defined NVIC_Set/GetVector with CMSIS implementation

Patater committed:
RTX5: uVisor: Use OsEventObserver

Patater committed:
RTX5: uVisor: Add OsEventObserver

Add the OsEventObserver mechanism. A client interested in receiving
notifications on certain OS events can register to receive notifications
with osRegisterForOsEvents. This is useful for clients like the secure
memory allocator, which observes thread switching events in order to
swap in and out different memory allocator objects.

Patater committed:
RTX5: uVisor: Extend thread control block with context

OsEventObserver objects expect a context to be maintained per thread on
their behalf. Add this context to the thread control block and extend
the thread creation functions with the ability to supply a context.

Patater committed:
RTX5: uVisor: Defer to uVisor for SVCall priority

Only set the SVCall priority if uVisor is not present. If uVisor is
present, keep using whatever priorities it has already set up.

Patater committed:
uVisor: Import RTX5-capable uVisor

Imported from 20170407_v7-M tag, commit e33f2739e961 "Make function in
transition macro match target".

Patater committed:
uVisor: Re-import RTX5-capable uVisor

Use a newer version of uVisor that doesn't change the box main thread
function type. Previously, we required all box main thread definitions
to change from taking a `const void *` to a `void *` when moving to
RTX5. We now are backwards compatibile.

c1728p9 committed:
Remove RTOS and CMSIS files

Remove the RTOS and CMSIS files in preperation for CMSIS 5. Note that
the RTOS heap ad stack test is left since this is still applicable to
RTX5. This file can be found here:
rtos\rtx\TARGET_CORTEX_M\TESTS\memory\heap_and_stack\main.cpp

Archcady committed:
Add up OS_MUTEX_NUM for ARMCC compiler

CI shield test (SPI test) may need 7 mutexes

bulislaw committed:
Update codebase for CMSIS5/RTX5

Update all of mbed-os to use RTX5.

bulislaw committed:
Pull in CMSIS 5 files

Pull in unmodified CMSIS5 files from the repository
https://github.com/ARM-software/CMSIS_5.git from the commit
c1b4232f4a98275dd79ebc99cf9b5da5053def4d.

The files were copied as shown below with source->dest_dir.
CMSIS_5/CMSIS/Core/Include/core_*.h -> mbed-os/cmsis/
CMSIS_5/CMSIS/Core_A/Include/core_*.h -> mbed-os/cmsis/
CMSIS_5/CMSIS/DSP/Include/arm_math.h -> mbed-os/cmsis/
CMSIS_5/CMSIS/Core/Include/tz_context.h -> mbed-os/cmsis/
CMSIS_5/CMSIS/Core/Include/cmsis_compiler.h -> mbed-os/cmsis/TARGET_CORTEX_M/
CMSIS_5/CMSIS/Core/Include/cmsis_arm*.h -> mbed-os/cmsis/TARGET_CORTEX_M/TOOLCHAIN_ARM/
CMSIS_5/CMSIS/Core/Include/cmsis_gcc*.h -> mbed-os/cmsis/TARGET_CORTEX_M/TOOLCHAIN_GCC/
CMSIS_5/CMSIS/Core_A/Include/cmsis_compiler.h -> mbed-os/cmsis/TARGET_CORTEX_A/
CMSIS_5/CMSIS/Core_A/Include/cmsis_arm*.h -> mbed-os/cmsis/TARGET_CORTEX_A/TOOLCHAIN_ARM/
CMSIS_5/CMSIS/RTOS2/RTX/Library/cmsis_os1.c -> mbed-os/rtos/rtx/
CMSIS_5/CMSIS/RTOS2/Include/cmsis_os2.h -> mbed-os/rtos/rtx2/TARGET_CORTEX_M/
CMSIS_5/CMSIS/RTOS2/RTX/Config/* -> mbed-os/rtos/rtx2/TARGET_CORTEX_M/
CMSIS_5/CMSIS/RTOS2/RTX/Include1/* -> mbed-os/rtos/rtx/
CMSIS_5/CMSIS/RTOS2/RTX/Include/* -> mbed-os/rtos/rtx2/TARGET_CORTEX_M/
CMSIS_5/CMSIS/RTOS2/RTX/Source/rtx_* -> mbed-os/rtos/rtx2/TARGET_CORTEX_M/
CMSIS_5/CMSIS/RTOS2/RTX/Source/core_cm.h -> mbed-os/rtos/rtx2/TARGET_CORTEX_M/
CMSIS_5/CMSIS/RTOS2/RTX/Source/ARM/irq_cm0.s -> mbed-os/rtos/rtx2/TARGET_CORTEX_M/TARGET_M0/TOOLCHAIN_ARM/
CMSIS_5/CMSIS/RTOS2/RTX/Source/GCC/irq_cm0.s -> mbed-os/rtos/rtx2/TARGET_CORTEX_M/TARGET_M0/TOOLCHAIN_GCC/
CMSIS_5/CMSIS/RTOS2/RTX/Source/IAR/irq_cm0.s -> mbed-os/rtos/rtx2/TARGET_CORTEX_M/TARGET_M0/TOOLCHAIN_IAR/
CMSIS_5/CMSIS/RTOS2/RTX/Source/ARM/irq_cm0.s -> mbed-os/rtos/rtx2/TARGET_CORTEX_M/TARGET_M0P/TOOLCHAIN_ARM/
CMSIS_5/CMSIS/RTOS2/RTX/Source/GCC/irq_cm0.s -> mbed-os/rtos/rtx2/TARGET_CORTEX_M/TARGET_M0P/TOOLCHAIN_GCC/
CMSIS_5/CMSIS/RTOS2/RTX/Source/IAR/irq_cm0.s -> mbed-os/rtos/rtx2/TARGET_CORTEX_M/TARGET_M0P/TOOLCHAIN_IAR/
CMSIS_5/CMSIS/RTOS2/RTX/Source/ARM/irq_cm3.s -> mbed-os/rtos/rtx2/TARGET_CORTEX_M/TARGET_M3/TOOLCHAIN_ARM/
CMSIS_5/CMSIS/RTOS2/RTX/Source/GCC/irq_cm3.s -> mbed-os/rtos/rtx2/TARGET_CORTEX_M/TARGET_M3/TOOLCHAIN_GCC/
CMSIS_5/CMSIS/RTOS2/RTX/Source/IAR/irq_cm3.s -> mbed-os/rtos/rtx2/TARGET_CORTEX_M/TARGET_M3/TOOLCHAIN_IAR/
CMSIS_5/CMSIS/RTOS2/RTX/Source/ARM/irq_cm4f.s -> mbed-os/rtos/rtx2/TARGET_CORTEX_M/TARGET_RTOS_M4_M7/TOOLCHAIN_ARM/
CMSIS_5/CMSIS/RTOS2/RTX/Source/GCC/irq_cm4f.s -> mbed-os/rtos/rtx2/TARGET_CORTEX_M/TARGET_RTOS_M4_M7/TOOLCHAIN_GCC/
CMSIS_5/CMSIS/RTOS2/RTX/Source/IAR/irq_cm4f.s -> mbed-os/rtos/rtx2/TARGET_CORTEX_M/TARGET_RTOS_M4_M7/TOOLCHAIN_IAR/

bulislaw committed:
Modify CMSIS 5 so it is suitable for mbed-os

Make changes to the files taken from RTX5/CMSIS5 so they work
with mbed-os and uvisor.

bridadan reviewed (changes requested):
@bulislaw There are still a mountain of warnings that are making the build times go through the roof for a number of targets: http://mbed-ci-master-2.austin.arm.com:8081/job/build_matrix/197/target=NUCLEO_F103RB,toolchain=GCC_ARM/console

That target went from taking ~10 seconds to ~9 minutes.

For your reference, here are the 3 targets that are problematic: 

```
9 min 4 sec, build_matrix/target=NUCLEO_L152RE,toolchain=GCC_ARM #197
9 min 3 sec, build_matrix/target=XDOT_L151CC,toolchain=GCC_ARM #197
8 min 47 sec, build_matrix/target=NUCLEO_F103RB,toolchain=GCC_ARM #197
```

bulislaw committed:
Remove redefinitions of register macros from target code

sg- reviewed (changes requested):
514d4bcdb46382e9db7d882b96f62e40807eae49 needs to be reverted. STACK_SIZES belong to an application, not a target.

If there is too much heap or stack, the tests should be able to run on devices with 16k RAM so may need to refactor tests or reduce STACK usage if just wrongly defined. Only MxF devices should change the required stack size

Commit:
Updated driver binaries for feature_cmsis5 as v2.0 rc1

bulislaw committed:
Make stack sizes configurable by the app

Application can overwrite stack sizes by defining main-stack-size and
thread-stack-size in mbed_app.json

sg- reviewed (changes requested):
Looking good!
- [ ] Expose name for Mutex in constructor?
- [ ] remote rtx_mbed.h and use symbols from linker / unified mbed boot?
- [ ] directory rename for cmsis/rtx
- [ ] OS uses all named threads and mutex or none. names should be more specific
- [ ] What is the difference now between the OS stack and main stack? Does changing the config for MBED_CONF_APP_MAIN_STACK_SIZE also change the config size of all default stack sizes?

bulislaw committed:
Rename directories rtx->rtx4 rtx2->rtx5

bulislaw committed:
Make sure all system threads and mutexes have clear names

bulislaw committed:
Fix IAR coding style and thread/mutex namingin mbed_boot.c

bulislaw committed:
Add name parameter for C++ mutex wrapper

bulislaw committed:
Remove deprecated RTX4 config options

bulislaw committed:
Make sure all RTOS attribute structures are 0-ed before use

bulislaw committed:
Add more verbose RTOS error messages

bridadan reviewed (approved):
Looks like you've managed to remove the mountain of build warnings! Thanks 👍

Commit:
Bump number of ARMC mutexes to fix PAL test failure

0xc0170 committed:
Jenkinsfile: remove client app test temporarily

This should be fixed as soon as client and its dependencies
are updated to resolve conflicts there. We expect this to be reverted asap.
